src/mpyl/stages/discovery.py

```diff
@@ -217,94 +205,160 @@ def find_projects_to_execute(
     changeset: Changeset,
     steps: Optional[StepsCollection],
 ) -> set[ProjectExecution]:
-    maybe_execution_projects = set(
-        map(
-            lambda project: _to_project_execution(
-                logger, project, stage, changeset, steps
+    def project_execution_if_needed(
+        project: Project,
+    ) -> Optional[ProjectExecution]:
+        if project.stages.for_stage(stage) is None:
+            return None
+
+        is_any_dependency_modified = any(
+            is_dependency_modified(logger, project, stage, changed_file, steps)
+            for changed_file in changeset.files_touched()
+        )
+        is_project_modified = any(
+            file_belongs_to_project(logger, project, changed_file)
+            for changed_file in changeset.files_touched()
+        )
+
+        if is_project_modified:
+            cache_key = _cache_key_from_changes_in_project(
+                logger=logger, project=project, changeset=changeset
+            )
+
+        elif is_any_dependency_modified:
+            cache_key = changeset.sha
+            logger.debug(
+                f"Project {project.name}: using git revision as cache key: {cache_key}"
+            )
```

Comment:
What's the benefit of this change? 

IMHO that would make this code a bit more confusing: currently it's very clear that we only try to calculate a hash if any of the project files was touched, otherwise we use the revision. With your proposed changes, one needs to dive into `_cache_key_from_changes_in_project()` (even when there are no changes to the project) just to learn what key we use in that case. Additionally, it would be harder to distinguish between the case when we are explicitly using the git revision as a key because only a dependency was touched) and the case when we are _falling back_ to the git revision because even though there were changes to this project, we cannot use them to calculate a hash (e.g. a file was deleted).

Reply:
It simply seemed like an extra check for the SHA for no reason, but it would make the code less readable in the end.

I do wonder if you should return the revision as a SHA anyway from this function. It's called `_cache_key_from_changes_in_project()` for a reason, but it falls back to the revision (which is unrelated to the changes) in case there are none? Feels to me like this function should return either a SHA from the changed files or None, so the we can then fall back to using the revision?

Reply:
That's actually a good point, I'll make that change 👍